### ISSUE TITLE
bug: dead loop while expand kvstore

### ISSUE THREAD

**Issue by ShawnHsiung** [buraksezer/olric#137]
### Issue description
1. put a big value to olricd,  and the size of value > tableSize
2. olricd want to expand the table, but if it's a idle table(or just created) which means the inuse=0, or (Inuse * 2) < len(value)
3. olricd go into a dead loop, which will cause the response timeout and node unavailable

internal/kvstore/kvstore.go
```
for {
    // Get the last value, storage only calls Put on the last created table.
    t := kv.tables[len(kv.tables)-1]
    err := t.put(hkey, value)
    if err == errNotEnoughSpace {
	    // Create a new table and put the new k/v pair in it.
	    nt := newTable(kv.Stats().Inuse * 2)
	    kv.tables = append(kv.tables, nt)
	    res = storage.ErrFragmented
	    // try again
	    continue
    }
    if err != nil {
	    return err
    }
    // everything is ok
    break
}
```

### Version
<= v0.4.1

**Comment by buraksezer** [comment]
Hi @ShawnHsiung 

I have just released [v0.4.2](https://github.com/buraksezer/olric/releases/tag/v0.4.2) to fix that problem. Thank you for the PR. I generalized the fix and added some tests. Here is my commit: https://github.com/buraksezer/olric/commit/23cd56d53a7d0998740e5a19204fb482003a5fbf

If you still have the same problem after upgrading, please re-open this issue.

**Comment by ShawnHsiung** [comment]
@buraksezer  the function **calculateTableSize** might have a little problem at a special situation
```
// minimum is the length of value, but actually there might be len(value)+len(key)+21
// assume 1: 
//     1. s.Inuse = 0
//     2. user defined minimumTableSize=tableSize < len(key)+21
// assume 2: 
//     1.  doubleInuse < minimumTableSize < (minimum +len(key)+21)

// so, the paramter **minimum** used may need replace with len(value)+len(key)+21 or len(value)+MaxKey+21

func (kv *KVStore) calculateTableSize(minimum int) int {
	s := kv.Stats()
	if s.Inuse == 0 { 
                // assume 1, still not enough
		// Value size is too big, and we are trying to allocate a new table to insert it into
		if kv.minimumTableSize <= minimum {
			return kv.minimumTableSize + minimum
		}

		// Fresh table
		return kv.minimumTableSize
	}
        
	doubledInuse := s.Inuse * 2

        // assume 2, still not enough
	// Minimum table size is kv.minimumTableSize
	if doubledInuse <= kv.minimumTableSize {
		return kv.minimumTableSize
	}
        // same with assume 2
	if doubledInuse <= minimum {
		return doubledInuse + minimum
	}
        // same with assume 2
	// Expand the table.
	return doubledInuse
}
```

**Comment by justinfx** [comment]
@buraksezer I just put my updated project into production, having switched from a simple in memory LRU to Olric, and I found a nasty variation of this bug that causes the memory to grow indefinitely until the machine runs out of memory. Here is a repro using v0.4.4. 

The key points are:

1.  We first put a key with a tiny value. Using 0 bytes in this example
2. We then put a key with a value that is too large for the table. Using 1MB in this example
3. The put call hangs while the memory allocations explode. Heap profile shows the code mostly in `newTable`.

```go
package main

import (
	"context"
	"log"
	"time"

	"github.com/buraksezer/olric"
	"github.com/buraksezer/olric/config"
)

func main() {
	c := config.New("local")

	ctx, cancel := context.WithCancel(context.Background())
	c.Started = func() {
		defer cancel()
		log.Println("[INFO] Olric is ready to accept connections")
	}

	db, err := olric.New(c)
	if err != nil {
		log.Fatalf("Failed to create Olric instance: %v", err)
	}

	go func() {
		err = db.Start()
		if err != nil {
			log.Fatalf("olric.Start returned an error: %v", err)
		}
	}()

	<-ctx.Done()

	dm, err := db.NewDMap("mybucket")
	if err != nil {
		log.Fatalf("olric.NewDMap returned an error: %v", err)
	}

	putGet := func(bval []byte) {
		log.Printf("Putting key with value size %d", len(bval))
		err = dm.Put("key", bval)
		if err != nil {
			log.Fatalf("Failed to call Put: %v", err)
		}
		log.Println("Getting value for key")
		val, err := dm.Get("key")
		if err != nil {
			log.Fatalf("Failed to call Get: %v", err)
		}
		bval = val.([]byte)
		log.Printf("Got value size: %d", len(bval))
	}

	log.Println("##\nOperations on a DMap instance:")

	// First put a key with an empty value.
	putGet([]byte{})
	// Then put a key with a value that is too large for the table.
	// WARNING: This is going to grow memory endlessly.
	const BadValueSize = 1 << 20
	putGet(make([]byte, BadValueSize))

	log.Println("##")

	// Don't forget the call Shutdown when you want to leave the cluster.
	ctx, _ = context.WithTimeout(context.Background(), 10*time.Second)
	err = db.Shutdown(ctx)
	if err != nil {
		log.Printf("Failed to shutdown Olric: %v", err)
	}
}

```

**Comment by buraksezer** [comment]
Hi @justinfx, thank you for reporting this.

I confirm the bug, and I have implemented a possible fix. The problem was in `calculateTableSize` function. It calculates a new table size when Olric needs more memory.

Here is my fix: 9ff76f76014553dbe121fbc19254b57fa7cc1548. I'm going to reconsider `calculateTableSize` function and add more tests. I assume that you are trying to cache large image files with Olric, right?

**Comment by justinfx** [comment]
Thanks for the fast reply! Yes you are correct. My application is an image cache service for preview quality images generated from an image conversion process. Usually the values are less than 1MB. But I ended up finding the problem because of a combination of factors.

- I implemented an "in flight" solution for multiple clients requesting the same image key, by having the first client put an empty record into the cache (hence the 0 byte value) before the image is processed, and then is replaced with the actual bytes.
- A user requested a preview image with a resolution higher than normal, which generated a value over 1MB

**Comment by justinfx** [comment]
@buraksezer I tested that patch and it looks like it does fix the problem, thanks! Are you planning to release a new 0.4.x patch or will this wait for your next 0.5.x?

**Comment by buraksezer** [comment]
Hi @justinfx 

I have just published `v0.4.5` to fix the problem you reported. If your average value size is around 1MB, you should consider picking a larger default table size. The default is 1048576 bytes. 

Programmatic configuration:

```go
c.StorageEngines.Config["kvstore"]["tableSize"] = 1000
```

Declerative configuration:

```yaml
storageEngines:
  config:
    kvstore:
      tableSize: 1048576
```

You can use `config.Load` function to configure your nodes in a declarative way. See `cmd/olricd/olricd.yaml` for sample configuration.

v0.5.x is a big step forward. It uses an improved version of the previous storage engine and it also uses the Redis protocol for communication. Documenting v0.5 is an ongoing effort but you can take a look at examples with the Golang client: https://github.com/buraksezer/olric#samples

List of the DMap commands: https://github.com/buraksezer/olric#distributed-map

Olric v0.5 also implements the Redis Publish-Subscribe service as a drop-in replacement.

@ShawnHsiung I also fixed the case you mentioned [here](https://github.com/buraksezer/olric/issues/137#issuecomment-1002386859).

**Comment by justinfx** [comment]
Epic! Thank you @buraksezer ! Both for the quick patch and the extra info about table size configuration!
